homeassistant/components/geniushub/climate.py

```diff
@@ -0,0 +1,159 @@
+    'footprint': STATE_ECO,
+    'away': None,
+    'override': STATE_MANUAL,
+    'early': STATE_HEAT,
+    'test': None,
+    'linked': None,
+    'other': None,
+}  # intentionally missing 'off': None
+GH_DEVICE_STATE_ATTRS = ['temperature', 'type', 'occupied', 'override']
+
+
+async def async_setup_platform(hass, hass_config, async_add_entities,
+                               discovery_info=None):
+    """Set up the Genius hub climate devices."""
+    client = hass.data[DOMAIN]['client']
+
+    zones = []
+    for zone in client.hub.zone_objs:
+        if hasattr(zone, 'temperature'):
+            zones.append(GeniusClimate(client, zone))
+
+    async_add_entities(zones, update_before_add=False)
+
+
+class GeniusClimate(ClimateDevice):
+    """Representation of a Genius Hub climate device."""
+
+    def __init__(self, client, zone):
+        """Initialize the climate device."""
+        self._client = client
+        self._objref = zone
+        self._id = zone.id
+        self._name = zone.name
+
+        # Only some zones have movement detectors, which allows footprint mode
+        op_list = list(HA_OPMODE_TO_GH)
+        if not hasattr(self._objref, 'occupied'):
+            op_list.remove(STATE_ECO)
+        self._operation_list = op_list
+
+    @property
+    def name(self):
+        """Return the name of the climate device."""
+        return self._objref.name
+
+    @property
+    def device_state_attributes(self):
+        """Return the device state attributes of the evohome Climate device.
+
+        This is state data that is not available otherwise, due to the
+        restrictions placed upon ClimateDevice properties, etc. by HA.
+        """
+        tmp = self._objref.__dict__.items()
+        state = {k: v for k, v in tmp if k in GH_DEVICE_STATE_ATTRS}
+
+        return {'status': state}
+
+    @property
+    def current_temperature(self):
+        """Return the current temperature."""
+        return self._objref.temperature
+
+    @property
+    def target_temperature(self):
+        """Return the temperature we try to reach."""
+        return self._objref.setpoint
+
+    @property
+    def min_temp(self):
+        """Return max valid temperature that can be set."""
+        return GENIUSHUB_MIN_TEMP
+
+    @property
+    def max_temp(self):
+        """Return max valid temperature that can be set."""
+        return GENIUSHUB_MAX_TEMP
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        return TEMP_CELSIUS
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        return GENIUSHUB_SUPPORT_FLAGS
+
+    @property
+    def operation_list(self):
+        """Return the list of available operation modes."""
+        return self._operation_list
+
+    @property
+    def current_operation(self):
+        """Return the current operation mode."""
+        return GH_STATE_TO_HA.get(self._objref.mode, None)
```

Comment:
`None` is the default value returned if the key is missing in the dict.

Reply:
@MartinHjelmare Can you confirm that `current_operation` is not called when `is_on` returns False?  I ask because I removed `'off': None` from `GH_STATE_TO_HA`, but put this in as a backup.

Reply:
`current_operation` is called as part of the state attributes, regardless of state.

Reply:
Is having `operation_mode==None` OK?  Specifically, is it OK to not have it in `operation_list`?

```
2019-04-16 22:15:11 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=climate.kitchen, 
old_state=<state climate.kitchen=off; current_temperature=19.1, min_temp=4.0, max_temp=28.0, temperature=18.5, operation_mode=None, operation_list=['auto', 'manual'], status=type=radiator, temperature=19.120001, override=duration=0, setpoint=16, friendly_name=Kitchen, supported_features=4225 @ 2019-04-16T21:14:21.050197+00:00>, 
new_state=<state climate.kitchen=off; current_temperature=19.1, min_temp=4.0, max_temp=28.0, temperature=4, operation_mode=None, operation_list=['auto', 'manual'], status=type=radiator, temperature=19.120001, override=duration=0, setpoint=16, friendly_name=Kitchen, supported_features=4225 @ 2019-04-16T21:14:21.050197+00:00>>
```

Reply:
Depending on what `is_on` returns, if `current_operation` returns `None` it could lead to `state` being unknown. It's ok if there's nothing better to return.

Reply:
`is_on` will always return `True` or `False`, never `None`.